File: src/main/java/com/example/hangman/Hangman.java
package com.example.hangman;

import java.util.LinkedHashSet;
import java.util.List;
import java.util.Locale;
import java.util.Set;

/**
 * The domain model for one hangman round.
 *
 * <p>The model keeps the order in which letters were attempted. That makes the state easy to
 * inspect in a console or web client while keeping repeated guesses unambiguous.
 */
public final class Hangman {

  private final String secretWord;
  private final int incorrectGuessLimit;
  private final Set<Character> guessedLetters = new LinkedHashSet<>();
  private final Set<Character> incorrectGuesses = new LinkedHashSet<>();

  public Hangman(String secretWord, int incorrectGuessLimit) {
    if (secretWord == null || secretWord.isBlank()) {
      throw new IllegalArgumentException("The secret word must not be blank");
    }

    String normalizedWord = secretWord.strip().toUpperCase(Locale.ROOT);
    if (!normalizedWord.matches("[A-Z]+")) {
      throw new IllegalArgumentException("The secret word must contain only letters A-Z");
    }
    if (incorrectGuessLimit <= 0) {
      throw new IllegalArgumentException("The incorrect guess limit must be positive");
    }

    this.secretWord = normalizedWord;
    this.incorrectGuessLimit = incorrectGuessLimit;
  }

  /** Attempts one letter, returning an outcome without changing state for invalid or duplicate input. */
  public GuessResult guess(String letter) {
    ensureRoundIsInProgress();

    if (letter == null || letter.length() != 1 || !isAsciiLetter(letter.charAt(0))) {
      return GuessResult.InvalidGuess;
    }

    char normalizedLetter = Character.toUpperCase(letter.charAt(0));
    if (guessedLetters.contains(normalizedLetter)) {
      return GuessResult.DuplicateGuess;
    }

    guessedLetters.add(normalizedLetter);
    if (!secretWord.contains(String.valueOf(normalizedLetter))) {
      incorrectGuesses.add(normalizedLetter);
      return GuessResult.IncorrectGuess;
    }

    return GuessResult.CorrectGuess;
  }

  /** Compatibility with the kata's original C#-style method name. */
  public GuessResult Guess(String letter) {
    return guess(letter);
  }

  public String getSecretWord() {
    return secretWord;
  }

  public int getIncorrectGuessLimit() {
    return incorrectGuessLimit;
  }

  public int getGuessesLeft() {
    return incorrectGuessLimit - incorrectGuesses.size();
  }

  public List<String> getIncorrectGuesses() {
    return asStrings(incorrectGuesses);
  }

  public List<String> getGuessedLetters() {
    return asStrings(guessedLetters);
  }

  public String getMaskedWord() {
    StringBuilder maskedWord = new StringBuilder(secretWord.length());
    for (int index = 0; index < secretWord.length(); index++) {
      char letter = secretWord.charAt(index);
      maskedWord.append(guessedLetters.contains(letter) ? letter : '-');
    }
    return maskedWord.toString();
  }

  public GameStatus getStatus() {
    if (isWordRevealed()) {
      return GameStatus.GameWon;
    }
    if (incorrectGuesses.size() >= incorrectGuessLimit) {
      return GameStatus.GameLost;
    }
    return GameStatus.GameIsOn;
  }

  public boolean isInProgress() {
    return getStatus() == GameStatus.GameIsOn;
  }

  public boolean isGameInProgress() {
    return isInProgress();
  }

  public boolean isWon() {
    return getStatus() == GameStatus.GameWon;
  }

  public boolean isLost() {
    return getStatus() == GameStatus.GameLost;
  }

  @Override
  public String toString() {
    return "Status:%s;Word:%s;Guesses:%s;Guesses left:%d".formatted(
      getStatus(),
      getMaskedWord(),
      String.join(" ", getGuessedLetters()),
      getGuessesLeft()
    );
  }

  private void ensureRoundIsInProgress() {
    if (!isInProgress()) {
      throw new IllegalStateException("The hangman round is already over");
    }
  }

  private boolean isWordRevealed() {
    for (int index = 0; index < secretWord.length(); index++) {
      if (!guessedLetters.contains(secretWord.charAt(index))) {
        return false;
      }
    }
    return true;
  }

  private static boolean isAsciiLetter(char letter) {
    return letter >= 'A' && letter <= 'Z' || letter >= 'a' && letter <= 'z';
  }

  private static List<String> asStrings(Set<Character> letters) {
    return letters.stream().map(String::valueOf).toList();
  }
}

File: src/main/webapp/app/hangman/domain/Hangman.ts
export type GameState = 'in-progress' | 'won' | 'lost';

export type GuessResult =
  | { kind: 'correct'; letter: string }
  | { kind: 'incorrect'; letter: string }
  | { kind: 'duplicate'; letter: string }
  | { kind: 'invalid'; letter: string }
  | { kind: 'game-over'; letter: string };

const LETTER = /^[A-Z]$/;

/**
 * The game rules and state for a single round of Hangman.
 *
 * A round accepts only one alphabetic letter per guess. Guesses are normalized
 * to upper case so that the UI can treat keyboard and button input alike.
 */
export class Hangman {
  private readonly word: string;
  private readonly incorrectGuessLimit: number;
  private readonly correctLetters = new Set<string>();
  private readonly wrongLetters: string[] = [];

  constructor(secretWord: string, incorrectGuessLimit: number) {
    const normalizedWord = secretWord.trim().toUpperCase();

    if (!/^[A-Z]+$/.test(normalizedWord)) {
      throw new Error('The secret word must contain at least one letter from A to Z.');
    }

    if (!Number.isInteger(incorrectGuessLimit) || incorrectGuessLimit < 1) {
      throw new Error('The incorrect guess limit must be a positive integer.');
    }

    this.word = normalizedWord;
    this.incorrectGuessLimit = incorrectGuessLimit;
  }

  get secretWord(): string {
    return this.word;
  }

  get state(): GameState {
    if (this.hasGuessedEveryLetter) {
      return 'won';
    }

    if (this.wrongLetters.length >= this.incorrectGuessLimit) {
      return 'lost';
    }

    return 'in-progress';
  }

  get inProgress(): boolean {
    return this.state === 'in-progress';
  }

  get incorrectGuesses(): readonly string[] {
    return [...this.wrongLetters];
  }

  get correctGuesses(): readonly string[] {
    return [...this.correctLetters];
  }

  get guessedLetters(): readonly string[] {
    return [...this.correctLetters, ...this.wrongLetters];
  }

  get incorrectGuessCount(): number {
    return this.wrongLetters.length;
  }

  get guessesLeft(): number {
    return Math.max(0, this.incorrectGuessLimit - this.wrongLetters.length);
  }

  get maskedWord(): string {
    return [...this.word].map(letter => (this.correctLetters.has(letter) ? letter : '—')).join(' ');
  }

  guess(rawLetter: string): GuessResult {
    const letter = rawLetter.trim().toUpperCase();

    if (!LETTER.test(letter)) {
      return { kind: 'invalid', letter };
    }

    if (!this.inProgress) {
      return { kind: 'game-over', letter };
    }

    if (this.correctLetters.has(letter) || this.wrongLetters.includes(letter)) {
      return { kind: 'duplicate', letter };
    }

    if (this.word.includes(letter)) {
      this.correctLetters.add(letter);
      return { kind: 'correct', letter };
    }

    this.wrongLetters.push(letter);
    return { kind: 'incorrect', letter };
  }

  private get hasGuessedEveryLetter(): boolean {
    return [...new Set(this.word)].every(letter => this.correctLetters.has(letter));
  }
}
